queries/period_slippage.sql

```diff
@@ -310,70 +310,114 @@ incoming_and_outgoing_with_buffer_trades as (
     from buffer_trades
 ),
 final_token_balance_sheet as (
-    select solver_address,
-           solver_name,
-           sum(amount) token_imbalance_wei,
-           symbol,
-           token,
-           tx_hash
-    from incoming_and_outgoing_with_buffer_trades
-    group by symbol,
-             token,
-             solver_address,
-             solver_name,
-             tx_hash
+    select
+        solver_address,
+        solver_name,
+        sum(amount) token_imbalance_wei,
+        symbol,
+        token,
+        tx_hash,
+        date_trunc('minute', block_time) as minute
+    from
+        incoming_and_outgoing_with_buffer_trades
+    group by
+        symbol,
+        token,
+        solver_address,
+        solver_name,
+        tx_hash,
+        block_time
+    having
+        sum(amount) != 0
 ),
-
-exact_prices as (
+token_times as (
     select
-        token as contract_address,
-        decimals,
-        price
-    from final_token_balance_sheet
-    left join prices.usd
-        on contract_address = token
-        and minute = '{{EndTime}}'
+        date_trunc('hour', minute) as hour,
+        minute,
+        token
+    from
+        final_token_balance_sheet
+    group by
+        hour,
+        minute,
+        token
 ),
--- The approximate prices are based on the hourly median prices over the past day.
-approximate_prices as (
+token_times_hour as (
     select
-        contract_address,
-        decimals,
-        percentile_cont(0.5) WITHIN group (order by median_price) as price
-    from prices.prices_from_dex_data p
-    where hour > '{{EndTime}}'::timestamptz - interval '1 day'
-    and date(hour) = '{{EndTime}}'::timestamptz - interval '1 day'
-      -- Getting approximate prices only for tokens we do not have an exact price for.
-    and contract_address IN (select contract_address from exact_prices where price is null)
-    group by contract_address, decimals
-    having sum(sample_size) > 0
+        hour,
+        token
+    from
+        token_times
+    group by
+        hour,
+        token
 ),
-end_prices as (
-    select * from exact_prices where price is not null
-    union
-    select * from approximate_prices
+precise_prices as (
+    select
+        pusd.*
+    from
+        prices.usd pusd
+        inner join token_times tt on pusd.minute = tt.minute
```

Comment:
This way you don't have to create two intermediary subqueries (any single-use subquery saved in this monster is a win I'd say)

```suggestion
        inner join final_token_balance_sheet tt on pusd.minute = date_trunc('minute', tt.block_time)
```

Reply:
If am doing this here:
```
inner join final_token_balance_sheet tt on pusd.minute = date_trunc('minute', tt.block_time)
```
then I guess you want me to do:
```
inner join token_and_time_of_trading_hourly tt on musd.hour =date_trunc('hour', tt.block_time)
        and contract_address = token
```
but then I get an issue, if there are several different block_times that trunc to the same hour. Then I get too many entries. Not sure how to deal with it. That is why I was introducing 2 tables

But maybe I don't understand your comment

Reply:
I see, makes sense then (maybe comment somewhere)?